pywbem_mock/_dmtf_cim_schema.py

```diff
@@ -0,0 +1,426 @@
+    """
+    def __init__(self, schema_version, schema_root_dir, use_experimental=False,
+                 verbose=False):
+        """
+        The constructor downloads the schema if the `schema_root_dir` or the
+        schema zip file do not exist and installs it into the directory defined
+        by the `schema_root_dir` parameter.
+
+        If the schema zip file is already installed it simply sets the property
+        values so that the schema can be compiled.
+
+        If the schema zip file was previously downloaded but not expanded, it is
+        expanded (i.e. the mof... subdirectory is created and the MOF files
+        expanded into this directory.).
+
+        Parameters:
+
+          schema_version  (tuple of 3 integers (x, y, z):
+            Represents the DMTF CIM schema version m.n.u where:
+
+            * m is the DMTF CIM schema major version
+            * n is the DMTF CIM schema minor version
+            * u is the DMTF CIM schema update version
+
+            This MUST represent a DMTF CIM schema that is available from the
+            DMTF web site.
+
+          schema_root_dir (:term:`string`):
+            Directory into which the DMTF CIM schema is installed or will be
+            installed.
+
+          use_experimental (:class:`py:bool`):
+            If `True`, the experimintal version of the defined schema is
+            installed.
+
+            If `False`, (the default) the Final released version of the DMTF
+            is installed.
+
+          verbose (:class:`py:bool`):
+            If `True`, progress messages are output to stdout.
+
+        Raises:
+            ValueError: if the schema cannot be retrieved
+                from the DMTF web site.
+            TypeError:  if the `schema_version` is not
+                a valid tuple with 3 integer components
+
+        """
+        if not isinstance(schema_version, tuple):
+            raise TypeError('schema_version must be tuple not %s' %
+                            schema_version)
+        if len(schema_version) != 3:
+            raise TypeError('DMTF Schema must have 3 integer components, '
+                            '(major, minor, update) not %s' % schema_version)
+        schema_type = 'Experimental' if use_experimental else 'Final'
+        self._schema_version = schema_version
+        self._schema_root_dir = schema_root_dir
+
+        mof_dir = 'mof%s%s' % (schema_type, self.schema_version_str)
+        self._schema_mof_dir = os.path.join(self._schema_root_dir, mof_dir)
+
+        self._dmtf_schema_version = 'cim_schema_%s' % (self.schema_version_str)
+        self._mof_zip_bn = '%s%s-MOFs.zip' % (self._dmtf_schema_version,
+                                              schema_type)
+        self._mof_zip_url = \
+            'http://www.dmtf.org/standards/cim/cim_schema_v%s%s%s/%s' % \
+            (schema_version[0], schema_version[1], schema_version[2],
+             self._mof_zip_bn)
```

Comment:
I suggest externalizing "self._mof_zip_url" into a property, simply because it is interesting to know where the zip file came from. We may want to name the property "schema_zip_url", though, and accordingly the internal attribute.

Reply:
DONE including rename